backends/dpdk/dpdkArch.cpp

```diff
@@ -2116,137 +2127,182 @@ const IR::Node* SplitP4TableCommon::postorder(IR::MethodCallStatement* statement
-        decls->push_back(gen_apply(memberTable));
+        IR::IfStatement* ret = nullptr;
+        if (implementation == TableImplementation::ACTION_SELECTOR) {
+            ifBaseTableHit->push_back(new IR::IfStatement(t0stat, t1stat, nullptr));
+            ifBaseTableHit->push_back(t2stat);
+            ret = new IR::IfStatement(cond, new IR::BlockStatement(*ifBaseTableHit), nullptr);
+        } else if (implementation == TableImplementation::ACTION_PROFILE) {
+            ret = new IR::IfStatement(cond, t2stat, nullptr);
+        }
+        decls->push_back(ret);
         return new IR::BlockStatement(*decls);
     }
     return statement;
 }
 
-// assume the RemoveMiss pass is applied
+// assume the RemoveMiss and SimplifyControlFlow pass is applied
 const IR::Node* SplitP4TableCommon::postorder(IR::IfStatement* statement) {
     auto cond = statement->condition;
-    bool negated = false;
-    if (auto neg = cond->to<IR::LNot>()) {
-        // We handle !hit, which may have been created by the
-        // removal of miss
-        negated = true;
-        cond = neg->expr;
-    }
-
     if (!P4::TableApplySolver::isHit(cond, refMap, typeMap)) return statement;
     if (!cond->is<IR::Member>()) return statement;
-
     auto member = cond->to<IR::Member>();
     if (!member->expr->is<IR::MethodCallExpression>()) return statement;
     auto mce = member->expr->to<IR::MethodCallExpression>();
     auto mi = P4::MethodInstance::resolve(mce, refMap, typeMap);
 
-    auto apply_hit = [](cstring table, bool negated) {
+    auto apply_hit = [](cstring table) {
         IR::Expression* expr =
             new IR::Member(new IR::MethodCallExpression(new IR::Member(
                                new IR::PathExpression(table), IR::ID(IR::IApply::applyMethodName))),
                            "hit");
-        if (negated) expr = new IR::LNot(expr);
         return expr;
     };
 
     if (auto apply = mi->to<P4::ApplyMethod>()) {
         if (!apply->isTableApply()) return statement;
         auto table = apply->object->to<IR::P4Table>();
         if (match_tables.count(table->name) == 0) return statement;
-        // an action selector t.apply() is converted to
-        // t0.apply();  // base table
-        // t1.apply();  // group table
-        // t2.apply();  // member table
-        //
-        // if (!t.apply().hit) {
-        //   foo();
-        // }
-        // is equivalent to
-        // if (t0.apply().hit) {
-        //   if (t1.apply().hit) {
-        //      if (t2.apply().hit) {
-        //        ; }
-        //      else {
-        //        foo(); }
-        //   else {
-        //      foo(); }
-        // else {
-        //   foo(); } } }
         //
         // if (t.apply().hit) {
         //   foo();
         // }
         // is equivalent to
         // if (t0.apply().hit) {
-        //   if (t1.apply().hit) {
-        //      if (t2.apply().hit) {
+        //    if (group_id != initial_groupid) {
```

Comment:
maybe you could add an else branch to the top if

Reply:
For the "if" in line:2203? we are only replacing the condition in top if and adding some additional statements within the ifTrue block. If there is an else in the input P4, it will still remain there.